File: travel-agent-strands/src/agents/dining_agent.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT-0

"""Dining specialist tool — port of agents/dining-agent.ts.

Exports a single ``suggest_dining`` Strands tool. Restaurants are filtered by
meal type and proximity to a named attraction, then sorted by travel time.
"""

from __future__ import annotations

from typing import List, Literal, Optional

from strands import tool

from ..data.restaurants import restaurants
from ..data.types import DiningRecommendation


@tool
def suggest_dining(
    attraction_name: str,
    meal_type: Literal["breakfast", "lunch", "dinner"],
    cuisine_preference: Optional[str] = None,
) -> str:
    """Suggest restaurants near a planned attraction appropriate for the specified meal type. Returns options with cuisine, price range, and travel time from the attraction.

    Args:
        attraction_name: Current or next attraction to find dining near.
        meal_type: Type of meal — one of 'breakfast', 'lunch', 'dinner'.
        cuisine_preference: Optional preferred cuisine type (matched
            case-insensitively); preferred cuisines are sorted first.

    Returns:
        Formatted multi-line text listing restaurant options.
    """
    # 1. Filter restaurants that serve the specified meal_type
    meal_filtered = [r for r in restaurants if meal_type in r.meal_types]

    # 2. Filter restaurants that are near the specified attraction
    nearby_filtered = [
        r
        for r in meal_filtered
        if any(na["attractionName"] == attraction_name for na in r.near_attractions)
    ]

    # 3. Build recommendations with travel time from the attraction
    recommendations: List[DiningRecommendation] = []
    for r in nearby_filtered:
        near_entry = next(
            na for na in r.near_attractions if na["attractionName"] == attraction_name
        )
        reasons = [
            f"Serves {meal_type}",
            f"{near_entry['travelTimeMinutes']} minutes from {attraction_name}",
            f"{r.cuisine_type} cuisine",
        ]
        recommendations.append(
            DiningRecommendation(
                restaurant=r,
                travel_time_from_attraction=near_entry["travelTimeMinutes"],
                match_reason=", ".join(reasons),
            )
        )

    # 4 & 5. Cuisine grouping then per-group sort by travel time
    if cuisine_preference:
        pref_lower = cuisine_preference.lower()
        preferred = [
            rec
            for rec in recommendations
            if rec.restaurant.cuisine_type.lower() == pref_lower
        ]
        others = [
            rec
            for rec in recommendations
            if rec.restaurant.cuisine_type.lower() != pref_lower
        ]
        preferred.sort(key=lambda x: x.travel_time_from_attraction)
        others.sort(key=lambda x: x.travel_time_from_attraction)
        recommendations = preferred + others
    else:
        recommendations.sort(key=lambda x: x.travel_time_from_attraction)

    # 6. Build text content for the model
    if recommendations:
        lines = [
            f"{i + 1}. {rec.restaurant.name} ({rec.restaurant.cuisine_type}, "
            f"{rec.restaurant.price_range}) - {rec.travel_time_from_attraction} min "
            f"from {attraction_name}. {rec.match_reason}"
            for i, rec in enumerate(recommendations)
        ]
        return "\n".join(lines)
    return f"No restaurants found serving {meal_type} near {attraction_name}."

File: travel-agent/src/agents/dining-agent.ts
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import { Type, type Static } from "@sinclair/typebox";
import type { AgentTool, AgentToolResult } from "@mariozechner/pi-agent-core";
import type { DiningRecommendation } from "../data/types.js";
import { restaurants } from "../data/restaurants.js";

const DiningQuerySchema = Type.Object({
  attraction_name: Type.String({ description: "Current or next attraction to find dining near" }),
  meal_type: Type.Union([
    Type.Literal("breakfast"),
    Type.Literal("lunch"),
    Type.Literal("dinner"),
  ], { description: "Type of meal" }),
  cuisine_preference: Type.Optional(Type.String({ description: "Preferred cuisine type" })),
});

type DiningQuery = Static<typeof DiningQuerySchema>;

export function createDiningAgentTool(): AgentTool<typeof DiningQuerySchema, DiningRecommendation[]> {
  return {
    name: "suggest_dining",
    label: "Dining Suggestions",
    description:
      "Suggest restaurants near a planned attraction appropriate for the specified meal type. Returns options with cuisine, price range, and travel time from the attraction.",
    parameters: DiningQuerySchema,
    execute: async (toolCallId, args, signal, onUpdate) => {
      return suggestDining(args);
    },
  };
}

export function suggestDining(args: DiningQuery): AgentToolResult<DiningRecommendation[]> {
  const { attraction_name, meal_type, cuisine_preference } = args;

  // 1. Filter restaurants that serve the specified meal_type
  const mealFiltered = restaurants.filter((r) =>
    r.mealTypes.includes(meal_type)
  );

  // 2. Filter restaurants that are near the specified attraction
  const nearbyFiltered = mealFiltered.filter((r) =>
    r.nearAttractions.some((na) => na.attractionName === attraction_name)
  );

  // 3. Build recommendations with travel time from the attraction
  let recommendations: DiningRecommendation[] = nearbyFiltered.map((r) => {
    const nearEntry = r.nearAttractions.find(
      (na) => na.attractionName === attraction_name
    )!;

    const reasons: string[] = [];
    reasons.push(`Serves ${meal_type}`);
    reasons.push(`${nearEntry.travelTimeMinutes} minutes from ${attraction_name}`);
    reasons.push(`${r.cuisineType} cuisine`);

    return {
      restaurant: r,
      travelTimeFromAttraction: nearEntry.travelTimeMinutes,
      matchReason: reasons.join(", "),
    };
  });

  // 4. If cuisine_preference is provided, prioritize matching cuisine first
  if (cuisine_preference) {
    const preferred = recommendations.filter(
      (rec) =>
        rec.restaurant.cuisineType.toLowerCase() ===
        cuisine_preference.toLowerCase()
    );
    const others = recommendations.filter(
      (rec) =>
        rec.restaurant.cuisineType.toLowerCase() !==
        cuisine_preference.toLowerCase()
    );
    recommendations = [...preferred, ...others];
  }

  // 5. Sort by travel time (closest first), preserving cuisine preference grouping
  // Within each group (preferred/others), sort by travel time
  if (cuisine_preference) {
    const preferred = recommendations.filter(
      (rec) =>
        rec.restaurant.cuisineType.toLowerCase() ===
        cuisine_preference.toLowerCase()
    );
    const others = recommendations.filter(
      (rec) =>
        rec.restaurant.cuisineType.toLowerCase() !==
        cuisine_preference.toLowerCase()
    );
    preferred.sort((a, b) => a.travelTimeFromAttraction - b.travelTimeFromAttraction);
    others.sort((a, b) => a.travelTimeFromAttraction - b.travelTimeFromAttraction);
    recommendations = [...preferred, ...others];
  } else {
    recommendations.sort((a, b) => a.travelTimeFromAttraction - b.travelTimeFromAttraction);
  }

  // 6. Build text content for the model
  const textContent =
    recommendations.length > 0
      ? recommendations
          .map(
            (rec, i) =>
              `${i + 1}. ${rec.restaurant.name} (${rec.restaurant.cuisineType}, ${rec.restaurant.priceRange}) - ${rec.travelTimeFromAttraction} min from ${attraction_name}. ${rec.matchReason}`
          )
          .join("\n")
      : `No restaurants found serving ${meal_type} near ${attraction_name}.`;

  return {
    content: [{ type: "text", text: textContent }],
    details: recommendations,
  };
}
